File: java/103.java
// 103. Binary Tree Zigzag Level Order Traversal
// Given a binary tree, return the zigzag level order traversal of its nodes' values. (ie, from left to right, then right to left for the next level and alternate between).
//
// For example:
// Given binary tree [3,9,20,null,null,15,7],
//
//     3
//    / \
//   9  20
//     /  \
//    15   7
// return its zigzag level order traversal as:
//
// [
//   [3],
//   [20,9],
//   [15,7]
// ]
//
// Runtime: 1 ms, faster than 73.92% of Java online submissions for Binary Tree Zigzag Level Order Traversal.
// Memory Usage: 39.4 MB, less than 22.59% of Java online submissions for Binary Tree Zigzag Level Order Traversal.

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) return ans;
        Queue<TreeNode> queue = new LinkedList<>();
        TreeNode cur;
        queue.add(root);
        int counter = 0;
        while(!queue.isEmpty()) {
            List<Integer> tempList = new ArrayList<>();
            int len = queue.size();
            for (int i = 0; i < len; i++) {
                cur = queue.remove();
                tempList.add(cur.val);
                if (cur.left != null) queue.add(cur.left);
                if (cur.right != null) queue.add(cur.right);
            }
            counter++;
            if (counter % 2 == 0) Collections.reverse(tempList);
            ans.add(tempList);
        }
        return ans;
    }
}

File: go/103.go
// 103. Binary Tree Zigzag Level Order Traversal

// Given the root of a binary tree, return the zigzag level order traversal of its nodes' values. (i.e., from left to right, then right to left for the next level and alternate between).

 

// Example 1:


// Input: root = [3,9,20,null,null,15,7]
// Output: [[3],[20,9],[15,7]]
// Example 2:

// Input: root = [1]
// Output: [[1]]
// Example 3:

// Input: root = []
// Output: []
 

// Constraints:

// The number of nodes in the tree is in the range [0, 2000].
// -100 <= Node.val <= 100

// Runtime 0 ms Beats 100.00%
// Memory 4.57 MB Beats 67.17%

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
    output := [][]int{}
    if root == nil {
        return output
    }
    queue := []*TreeNode{}
    queue = append(queue, root)
    isEven := true
    for ; len(queue) > 0; {
        size := len(queue)
        row := []int{}
        for i := 0; i < size; i++ {
            cur := queue[i]
            row = append(row, cur.Val)
            if cur.Left != nil {
                queue = append(queue, cur.Left)
            }
            if cur.Right != nil {
                queue = append(queue, cur.Right)
            }
        }
        if isEven == false {
            slices.Reverse(row)
        }
        isEven = !isEven
        output = append(output, row)
        queue = queue[size:]
    }
    return output
}
